libcanard/canard.h

```diff
@@ -245,54 +232,111 @@ typedef struct
-/// subscription will be silently dropped by the library.
+/// subscription will be silently dropped by the library. The entire RX pipeline is invariant to the number of
+/// redundant CAN interfaces used.
 ///
-/// WARNING: SUBSCRIPTION INSTANCES SHALL NOT BE COPIED OR MUTATED BY THE APPLICATION (except user_reference).
+/// SUBSCRIPTION INSTANCES SHALL NOT BE MOVED WHILE IN USE.
 ///
 /// The memory footprint of a subscription is large. On a 32-bit platform it slightly exceeds half a KiB.
 /// This is an intentional time-memory trade-off: use a large look-up table to ensure predictable temporal properties.
 typedef struct CanardRxSubscription
 {
-    struct CanardRxSubscription* next;  ///< Read-only DO NOT MODIFY THIS
+    CanardTreeNode base;  ///< Read-only DO NOT MODIFY THIS
+
+    CanardMicrosecond transfer_id_timeout_usec;
+    size_t            extent;   ///< Read-only DO NOT MODIFY THIS
+    CanardPortID      port_id;  ///< Read-only DO NOT MODIFY THIS
+
+    /// This field can be arbitrarily mutated by the user. It is never accessed by the library.
+    /// Its purpose is to simplify integration with OOP interfaces.
+    void* user_reference;
 
     /// The current architecture is an acceptable middle ground between worst-case execution time and memory
     /// consumption. Instead of statically pre-allocating a dedicated RX session for each remote node-ID here in
     /// this table, we only keep pointers, which are NULL by default, populating a new RX session dynamically
-    /// on an ad-hoc basis when we first receive a transfer from that node. This is deterministic because our memory
-    /// allocation routines are assumed to be deterministic and we make at most one allocation per remote node,
-    /// but the disadvantage is that these additional operations lift the upper bound on the execution time.
-    /// Further, the pointers here add an extra indirection, which is bad for systems that leverage cached memory,
-    /// plus a pointer itself takes about 2-8 bytes of memory, too.
+    /// on an ad-hoc basis when we first receive a transfer from that node. This is O(1) because our memory
+    /// allocation routines are assumed to be O(1) and we make at most one allocation per remote node.
     ///
-    /// A far more predictable and a much simpler approach is to pre-allocate states here statically instead of keeping
+    /// A more predictable and simpler approach is to pre-allocate states here statically instead of keeping
     /// just pointers, but it would push the size of this instance from about 0.5 KiB to ~3 KiB for a typical 32-bit
     /// system. Since this is a general-purpose library, we have to pick a middle ground so we use the more complex
     /// but more memory-efficient approach.
-    struct CanardInternalRxSession* _sessions[CANARD_NODE_ID_MAX + 1U];
+    struct CanardInternalRxSession* sessions[CANARD_NODE_ID_MAX + 1U];  ///< Read-only DO NOT MODIFY THIS
```

Comment:
Regarding @thirtytwobits statement
> I'll keep reading but why are we giving up on 4k RAM?

Would utilizing CAVL here save enough memory to support 4KB RAM targets?
The maximum number of sessions is 128 thus would give you at most 7 hops in a binary tree.
Didn't look into the specifics of CAVL yet, but I suppose it's more memory efficient then allocating 128 pointers.

Reply:
>The maximum number of sessions is 128 thus would give you at most 7 hops in a binary tree.

10 hops in the worst case (the upper bound is a bit more involved than log2(size))

>Didn't look into the specifics of CAVL yet, but I suppose it's more memory efficient then allocating 128 pointers.

This is true. From the time complexity perspective though, CAVL is less desirable because O(1) array lookup is much easier to analyze and validate than O(log n) tree search. We can certainly go this way but should we really?